utils/src/lib.rs

```diff
@@ -62,9 +65,41 @@ pub fn remove_extra_costs(gas: U256, calldata: &[u8]) -> U256 {
     gas - calldata_cost - BASE_TX_COST
 }
 
+/// Flattens a group of contracts into maps of all events and functions
+pub fn flatten_funcs_and_events(
+    contracts: &BTreeMap<String, (Abi, Vec<u8>)>,
+) -> (BTreeMap<[u8; 4], Function>, BTreeMap<H256, Event>, Abi) {
+    let flattened_funcs: BTreeMap<[u8; 4], Function> = contracts
+        .iter()
+        .flat_map(|(_name, (abi, _code))| abi.functions().cloned().collect::<Vec<Function>>())
+        .collect::<Vec<Function>>()
+        .into_iter()
+        .map(|func| (func.short_signature(), func))
+        .collect();
+
+    let flattened_events: BTreeMap<H256, Event> = contracts
+        .iter()
+        .flat_map(|(_name, (abi, _code))| abi.events().cloned().collect::<Vec<Event>>())
+        .collect::<Vec<Event>>()
+        .into_iter()
+        .map(|event| (event.signature(), event))
+        .collect();
+
+    // We need this for better revert decoding, and want it in abi form
+    let mut errors_abi = Abi::default();
+    contracts.iter().for_each(|(_name, (abi, _code))| {
+        abi.errors().for_each(|error| {
+            let entry =
+                errors_abi.errors.entry(error.name.clone()).or_insert_with(Default::default);
+            entry.push(error.clone());
+        });
+    });
```

Comment:
This should also be doable with `.map()` / `.flatten()` I think

Reply:
Can you do that when sometimes you need to push to the inner vector? Wont a `map` style implementation overwrite some values?

Reply:
Hmm maybe you're right. Probably not a problem anyway, this should be fast enough